### ISSUE TITLE
IPython notebook for 10 fold cross-validation with ERT's?

### ISSUE THREAD

**Issue by MechCoder** [mjbommar/scotus-predict#4]
Hi,

I quickly skimmed through the paper and it states that 10-fold cross-validation is applied to fix the parameters `min_samples_leaf`, `max_depth`, `max_features`, `n-estimators`. But it seems to me that in the IPython notebooks in the repo, for example (https://github.com/mjbommar/scotus-predict/blob/master/build_scdb_model_20_extra_trees.ipynb) the values of `n_estimators` keep increasing for the fit in every term and the other values are not changed from the default values.

It would be great if you could point me to the part of the repository that does the cross-validation.

Thanks!!

**Comment by mjbommar** [comment]
We've moved on to a different model since the original publication.  You can find that branch (and the specific CV setup) here:
https://github.com/mjbommar/scotus-predict/blob/model-201408/src/scotus/model.py#L81

As you noted, the new implementation is a growing tree ensemble with an optional, periodic-or-event-driven forest fire.

**Comment by mjbommar** [comment]
Also, for sake of reproduction, we had fixed the parameter search space in the repo so people didn't end up with a 20 day run.

**Comment by MechCoder** [comment]
I see. In the new "growing tree ensemble" were you able to beat the 70.9% per-justice outcome and 69.7% per-case outcome accuracy? Where are these new numbers reported?

Thanks!

**Comment by mjbommar** [comment]
The original binary objective measure created some complications when we
mapped the individual votes into the aggregate outcome, so the newer models
are trained with an OvR 3 option combo.  Under this setup, the expanding
forest outperformed the CV'd ERT model and has much faster training times.
We have also tested a few DNN approaches but, in rigorous back testing,
haven't seen any improvement on these ensemble method despite the
substantial increase in training time and GPU requirement.

Our current efforts have shifted towards new unstructured feature sources,
but we'd be interested in looking at the original model again if you have
any thoughts.

Thanks,
Michael J. Bommarito II, CEO
Bommarito Consulting, LLC
Web: http://www.bommaritollc.com
Mobile: +1 (646) 450-3387

**Comment by MechCoder** [comment]
Thanks. A few more questions which I would be glad if you have the time to answer.
1. The script with the best model is (https://github.com/mjbommar/scotus-predict/blob/master/build_scdb_model_20_extra_trees.ipynb) right?  Going by the code, the accuracy for the per-justice outcome that is displayed that is 0.654 is for the last term only, but the accuracy averaged across all the terms betters the previous model? On a side-note, you should fix the random_state parameter in the script for the same feature splits for reproducabilty.
2.  Could you elaborate a bit on what the complications were? I don't have enough domain knowledge on this but what does it mean in reality when the labels are -1. Does it mean that the decision is still pending in court or something similar? Do you think just removing the samples corresponding to these labels will improve the accuracy,  simply because we are interested only in predicting affirm or reverse decisions and don't really care about whether the case is still pending or not?
3. In the old model, if I am understanding correctly, using a StratifiedKFold is not the best way to cross-validate time series, because let us say your training data has years 1953, 1954 and 1955, if a StratifiedKFold would fit on some samples in the data corresponding to 1955 and validate it on the previous year. We are trying to provide CV objects for this use-case here https://github.com/scikit-learn/scikit-learn/pull/6351
4. In the new model, you fix the parameters to be default right? Or is that also a reproduction concern? Do you think it would be worth investing efforts to do a time-series cross-validation here?

This is for a course-project. We are trying to improve the state-of-the-art by trying to add more predictive features. But for that we would also like to know if the existing model is the best as it can get.

**Comment by MechCoder** [comment]
cc: @pyeres @vpm238

**Comment by mjbommar** [comment]
Quick response is that I am very very interested in your walk-forward/time
series cross-validation suggestion in #3.  I had proposed this a few years
ago but had it shot down (as usual) by Gael.

Thanks,
Michael J. Bommarito II, CEO
Bommarito Consulting, LLC
_Web:_ http://www.bommaritollc.com
_Mobile:_ +1 (646) 450-3387

**Comment by mjbommar** [comment]
https://github.com/scikit-learn/scikit-learn/issues/3202

Thanks,
Michael J. Bommarito II, CEO
Bommarito Consulting, LLC
_Web:_ http://www.bommaritollc.com
_Mobile:_ +1 (646) 450-3387

**Comment by mjbommar** [comment]
And if you can email me directly (mike@lexpredict.com or
michael@bommaritollc.com), we can discuss some of the new features source
we've evaluated as well.

Thanks,
Michael J. Bommarito II, CEO
Bommarito Consulting, LLC
_Web:_ http://www.bommaritollc.com
_Mobile:_ +1 (646) 450-3387
